modules/aws/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/aws/KinesisProducer.scala

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2021-2023 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.rdbloader.aws
+
+import cats.effect.concurrent.Ref
+import cats.effect.{Blocker, ContextShift, Resource, Sync, Timer}
+import cats.implicits._
+import cats.{Applicative, Monoid, Parallel}
+import com.snowplowanalytics.snowplow.rdbloader.common.cloud.Queue
+import com.snowplowanalytics.snowplow.rdbloader.common.config.Region
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import retry.syntax.all._
+import retry.{RetryPolicies, RetryPolicy}
+import software.amazon.awssdk.core.SdkBytes
+import software.amazon.awssdk.regions.{Region => AWSRegion}
+import software.amazon.awssdk.services.kinesis.KinesisClient
+import software.amazon.awssdk.services.kinesis.model.{PutRecordsRequest, PutRecordsRequestEntry, PutRecordsResponse}
+
+import java.net.URI
+import java.nio.charset.StandardCharsets.UTF_8
+import java.util.UUID
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.FiniteDuration
+
+object KinesisProducer {
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  def producer[F[_]: ContextShift: Parallel: Sync: Timer](
+    streamName: String,
+    region: Region,
+    customEndpoint: Option[URI],
+    blocker: Blocker,
+    internalErrorsPolicy: BackoffPolicy,
+    throttlingPolicy: BackoffPolicy,
+    requestLimits: RequestLimits
+  ): Resource[F, Queue.ChunkProducer[F]] =
+    Resource
+      .fromAutoCloseable(Sync[F].delay(buildKinesisClient(customEndpoint, region)))
+      .map { client =>
+        new Queue.ChunkProducer[F] {
+          override def send(groupId: Option[String], messages: List[String]): F[Unit] = {
+            val records = messages.map(toKinesisRecord)
```

Comment:
What about using the `groupId` as a prefix in the partition key ? Maybe it could help for some troubleshooting

Reply:
So it looks like `groupId` is constant in the streaming transformer and is defined as `shredding` in `Completion` file.
It also looks like it's only needed in the context of AWS queue-like resources: SQS and SNS. It's not used for Pubsub. I don't think it would be useful in troubleshooting.

Maybe we should get rid of `groupId` from this interface? If it's constant and specific for SQS/SNS, we should pass groupId in constructor when building resources? WDYT?

Reply:
> So it looks like groupId is constant in the streaming transformer

It is constant for one instance, but is not shared across transformer instances, right? If that's the case I still think that it would be valuable to add it, as it could help to identify if all the bad rows come from a same EMR node. It might never be useful, but it's good to provide as much info as we can for troubleshooting.

> Maybe we should get rid of groupId from this interface? If it's constant and specific for SQS/SNS, we should pass groupId in constructor when building resources? WDYT?

If we don't put it in the partition key I agree!

Reply:
It's hardcoded with value `shredding` [here](https://github.com/snowplow/snowplow-rdb-loader/blob/50e72b0cd24012a573a6ec5f0830f8c426fd99d6/modules/common-transformer-stream/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/stream/common/Completion.scala#L37). It's the same for every instance.

Reply:
Ah right, I thought that it was the id that each instance uses in the folder path